### PR TITLE
Multivariate T

### PR DESCRIPTION
**Pull request by TMiguelT**
* Introduces a pluggable prior/posterior config for multivariate Gaussian data, with sensible defaults. Note that this only works for scipy > 1.6.0, where they introduced the multivariate t PDF. The library will remind you to upgrade if you have an old version.
* Adds a test for this new configuration, as well as for the univariate one
* Adds a "dev" and "multivariate" setup extra, meaning that you can `pip install bayesian_changepoint_detection[dev]` for development work (currently this installs pytest), or `pip install bayesian_changepoint_detection[multivariate]` (enforces that you have a new enough scipy version for this new feature)

### CODE REVIEW

File: bayesian_changepoint_detection/online_changepoint_detection.py [hunk 1]
```diff
@@ -65,3 +68,80 @@ def update_theta(self, data):
...
+
+class MultivariateT:
+    def __init__(self, dims, dof=None, kappa=1, mu=None, scale=None, chunksize=1):
+        """
+        Create a new predictor using the multivariate student T distribution as the posterior.
+            This implies a multivariate Gaussian distribution on the data, a Wishart prior on the precision,
+             and a Gaussian prior on the mean.
+             Implementation based on Haines, T.S., Gaussian Conjugate Prior Cheat Sheet.
+        :param dof: The degrees of freedom on the prior distribution of the precision (inverse covariance)
+        :param kappa: TODO
```
Comment (hildensia) [inline bayesian_changepoint_detection/online_changepoint_detection.py:80]: finish TODO

File: setup.py [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
-setup(name='bayesian_changepoint_detection',
-      version=bayesian_changepoint_detection.__version__,
-      description='Some Bayesian changepoint detection algorithms',
-      author='Johannes Kulick',
-      author_email='johannes.kulick@ipvs.uni-stuttgart.de',
-      url='http://github.com/hildensia/bayesian_changepoint_detection',
-      packages = ['bayesian_changepoint_detection'],
-      requires=['scipy', 'numpy']
-     )
+setup(
```
Comment (hildensia) [inline setup.py:7]: Nice, thanks!

File: test.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+def test_multivariate():
+    # 10-dimensional multivariate normal
+    distr = multivariate_normal(mean=[0]*10)
+    dataset = np.vstack((multivariate_normal.rvs([0]*10, size=50), multivariate_normal.rvs([2]*10, size=50)))
+    r, maxes = online.online_changepoint_detection(
+        dataset,
+        partial(offline.const_prior, l=(len(dataset) + 1)),
+        online.MultivariateT(10)
+    )
+    # TODO: assertion
```
Comment (hildensia) [inline test.py:17]: One possibility: just do a snapshot test here

Comment (TMiguelT) [reply test.py]: I'm not sure what you mean by this. Still, I've added an assertion based on https://github.com/hildensia/bayesian_changepoint_detection/issues/18#issuecomment-640391877

**Comment by TMiguelT** [comment]
Oops, a colleague just pointed out a huge flaw in my implementation. I'm putting this back to draft while I fix it.

File: bayesian_changepoint_detection/online_changepoint_detection.py [hunk 4]
```diff
@@ -60,8 +64,100 @@ def update_theta(self, data):
...
+    def update_theta(self, data):
+        """
+        Performs a bayesian update on the prior parameters, given data
+        :param data: A 1 x D vector of new data
+        """
+        if self.n >= self.kappa.shape[0]:
+            self.expand()
+        centered = (data - self.mu[self.n - 1])
+
+        self.dof[self.n] = self.dof[self.n - 1] + 1
```
Comment (TMiguelT) [inline bayesian_changepoint_detection/online_changepoint_detection.py:160]: Each `param[t]` doesn't depend on `param[t-1]`. Rather, each `param[t, r]` depends on `param[t-1, r-1]`. So I should copy the existing univariate behaviour which updates the entire vector simultaneously as we increase `t`.

File: test.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from scipy.stats import multivariate_normal, norm
+from functools import partial
+import numpy as np
+import bayesian_changepoint_detection.online_changepoint_detection as online
+
+
+def test_multivariate():
```
Comment (TMiguelT) [inline test.py:7]: This test data needs to have multiple breakpoints. Having only one hid the flaw in my implementation, which is that it wasn't updating parameters except for `r=0 and r=t`.

**Comment by TMiguelT** [comment]
Well, I don't love the accuracy of it, but does seem to output a meaningful signal for multiple changepoints. I also copied a plot function used in the notebook into the module for others' use. Here is an example calculated using the multivariate test data:  
image

File: bayesian_changepoint_detection/online_changepoint_detection.py [hunk 6]
```diff
@@ -60,8 +64,108 @@ def update_theta(self, data):
...
+        :param mu: The mean of the prior distribution on the mean
+        :param scale: The mean of the prior distribution on the precision
+        :param dims: The number of variables
+        """
+        # We default to the minimum possible degrees of freedom, which is 1 greater than the dimensionality
+        if dof is None:
+            dof = dims + 1
+        # The default mean is all 0s
+        if mu is None:
+            mu = [0]*dims
```
Comment (hildensia) [inline bayesian_changepoint_detection/online_changepoint_detection.py:91]: nitpick: spaces around `*`

**Review by hildensia — APPROVED** [review]

**Pull request merged**